### ISSUE TITLE
Runtime Errors and Warnings with INSCameraSDKSample-3.0.4_403

### ISSUE THREAD

**Issue by shoaoki-arent** [Insta360Develop/CameraSDK-iOS#40]
I have integrated the Framework from what appears to be the latest SDK version, INSCameraSDKSample-3.0.4_403, into my project. Upon building, I encountered the following issues. Could these issues be specific to my environment?

1. Runtime Log Warnings
When including INSCameraSDK.xcframework and INSCoreMedia.framework in my project and setting both to Embed & Sign under Frameworks, Libraries, and Embedded Content, the following warning appears in the log:

```
objc[1700]: Class INSPBWindowCropInfoRoot is implemented in both /private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/INSCameraSDK.framework/INSCameraSDK (0x1048645e0) and /private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/INSCoreMedia.framework/INSCoreMedia (0x10e0a9300). One of the two will be used. Which one is undefined.
objc[1700]: Class INSPBWindowCropInfo is implemented in both /private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/INSCameraSDK.framework/INSCameraSDK (0x104864608) and /private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/INSCoreMedia.framework/INSCoreMedia (0x10e0a9328). One of the two will be used. Which one is undefined.
objc[1700]: Class INSPBOffsetStateRoot is implemented in both /private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/INSCameraSDK.framework/INSCameraSDK (0x104864860) and /private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/INSCoreMedia.framework/INSCoreMedia (0x10e0a9d50). One of the two will be used. Which one is undefined.
```
Previously, I used INSCameraSDK.xcframework and INSCoreMedia.framework from INSCameraSDK-SampleOC-2.9.43_388, and these warnings were not present. Also, INSCoreMedia is a framework, not an xcframework.

2. Runtime Errors with xcframework
Using INSCameraSDK.xcframework and INSCoreMedia.xcframework results in runtime errors with the following log:
```
dyld[1622]: Library not loaded: @rpath/SSZipArchive.framework/SSZipArchive
  Referenced from: <6AAA3B66-16B8-35E0-892F-16377AF003A2> /private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/INSCoreMedia.framework/INSCoreMedia
  Reason: tried: '/private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/SSZipArchive.framework/SSZipArchive' (no such file), '/private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/INSCoreMedia.framework/Frameworks/SSZipArchive.framework/SSZipArchive' (no such file), '/private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/SSZipArchive.framework/SSZipArchive' (no such file), '/private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/INSCameraSDK.framework/Frameworks/SSZipArchive.framework/SSZipArchive' (no such file), '/usr/lib/swift/SSZipArchive.framework/SSZipArchive' (no such file, not in dyld cache), '/private/preboot/Cryptexes/OS/usr/lib/swift/SSZipArchive.framework/SSZipArchive' (no such file), '/private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/SSZipArchive.framework/SSZipArchive' (no such file), '/private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/SSZipArchive.framework/SSZipArchive' (no such file), '/usr/lib/swift/SSZipArchive.framework/SSZipArchive' (no such file, not in dyld cache), '/private/preboot/Cryptexes/OS/usr/lib/swift/SSZipArchive.framework/SSZipArchive' (no such file), '/private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/SSZipArchive.framework/SSZipArchive' (no such file), '/private/var/containers/Bundle/Application/{UUID}/{my app name}.app/Frameworks/SSZipArchive.framework/SSZipArchive' (no such file), '/System/Library/Frameworks/SSZipArchive.framework/SSZipArchive' (no such file, not in dyld cache)

```
Both are set to Embed & Sign in Frameworks, Libraries, and Embedded Content, and both are using xcframework. As a preliminary fix, I tried adding SSZipArchive via pod install, which then indicated that NvEffectSdkCore was required.

Thank you for your assistance!

**Comment by sssuourabh** [comment]
Hi are you able to compile? I am not able to compile even.
<img width="379" alt="Screenshot 2024-06-20 at 10 06 49 AM" src="https://github.com/Insta360Develop/CameraSDK-iOS/assets/2918573/6242282e-b39e-4281-8e95-08a47900d878">

**Comment by sssuourabh** [comment]
this is the output i am getting on Xcode 15 .

**Comment by sssuourabh** [comment]
Please ignore my comments I am able to successfully run the example app.

**Comment by shoaoki-arent** [comment]
@sssuourabh 
Thank you for your comment.
I found it difficult to run the sample app, so I decided not to proceed with it.
I already have an older version of the INSCameraSDK embedded in my app, so I thought it would be sufficient to just replace it.

By the way, the sample app seems to be embedding a `.framework` instead of an `.xcframework`, so I don't think it provides any hints for solving the second point in my comment.

**Comment by sssuourabh** [comment]
@shoaoki-arent actually i figured out how to run the sample app. I haven't used in my app till now. if you need any help on running the sample app drop me a message. Thanks

**Comment by shoaoki-arent** [comment]
@sssuourabh 
Thank you very much.

When you have some time, could you please let me know what you did to get the sample app running? I'm sure it will be helpful to others as well.

Here’s what I have done so far:
I tried fixing the path to the framework in the Frameworks directory (I couldn't find NSLogger.framework).
![image](https://github.com/Insta360Develop/CameraSDK-iOS/assets/155700565/a33008d2-d13b-4e7f-9621-64cb966af7bc)

After that, I'm encountering the following error with Eureka. It might be because I am using Xcode 15.2.
However, this is the result of testing on a simulator, and the result may be different on a real device.
```
Compiled module was created by a different version of the compiler '5.9.0.128.108'; rebuild 'Eureka' and try again: /Users/{username}/Library/Developer/Xcode/DerivedData/INSCameraSDKSample-bluetooth-bnxikozjplyitngpcnkyktrpwnet/Index.noindex/Build/Products/Debug-iphoneos/Eureka.framework/Modules/Eureka.swiftmodule/arm64-apple-ios.swiftmodule
```

**Comment by sssuourabh** [comment]
Hi, Please follow these steps -

**Comment by ravilich86** [comment]
Hi @shoaoki-arent! I also have Xcode 15.2. As @sssuourabh suggested, you need to update the Cartfile. I also noticed that you have both INSCoreMedia.xcframework and INSCoreMedia.framework. You only need one of them. I removed NSLogger.framework entirely, and everything compiles for me.
<img width="314" alt="Снимок экрана 2024-06-22 в 15 30 24" src="https://github.com/Insta360Develop/CameraSDK-iOS/assets/570642/ca9f4333-8ae6-4cd6-8bd0-baad555dbe51">
<img width="301" alt="Снимок экрана 2024-06-22 в 15 30 32" src="https://github.com/Insta360Develop/CameraSDK-iOS/assets/570642/3f1d6bb3-a1c9-48f3-bcba-df9e4528ddeb">

**Comment by shoaoki-arent** [comment]
@sssuourabh @ravilich86 
Thank you very much. I appreciate both of your assistance.
I successfully managed to build on the actual device.
I was struggling with errors, but after extracting the sample project again and starting over from the beginning, it worked out perfectly.
Thank you for the wonderful tip!

**Comment by shoaoki-arent** [comment]
P.S.
I understand that it is pointless since we are working with an real camera, but when trying to build on the simulator, I encountered an error in INSCameraSDK.xcframework indicating something like "no build for the simulator." Therefore, I have given up on running it on the simulator.

**Comment by jessearmandse** [comment]
@shoaoki-arent To have a universal framework, you'd need to build another simulator target. This can be an empty target in Xcode with sdk set to `iphonesimulator`. Then, you'd have to create a new `xcframework` from both architecture.

The resulting `xcframework` should contain these files:
```
Info.plist
ios-arm64
ios-arm64_x86_64-simulator
```
Although I wish Insta360 would just provide a proper `INSCameraSDK.xcframework` and `INSCoreMedia.xcframework` with universal architecture for us 😄

**Comment by shoaoki-arent** [comment]
@jessearmandse 
Thank you for the information.
Does it actually work in the simulator, or is it solely intended to ensure a successful build?

**Comment by jessearmandse** [comment]
It's just to ensure a successful build. I guess you could also just link it on Apple Silicon machine, and bypass the simulator entirely.

**Comment by sssuourabh** [comment]
hi @jessearmandse can you please provide more info on this - https://github.com/Insta360Develop/CameraSDK-iOS/issues/40#issuecomment-2185851367

@shoaoki-arent

**Comment by sssuourabh** [comment]
> @shoaoki-arent To have a universal framework, you'd need to build another simulator target. This can be an empty target in Xcode with sdk set to `iphonesimulator`. Then, you'd have to create a new `xcframework` from both architecture.
> 
> The resulting `xcframework` should contain these files:
> 
> ```
> Info.plist
> ios-arm64
> ios-arm64_x86_64-simulator
> ```
> 
> Although I wish Insta360 would just provide a proper `INSCameraSDK.xcframework` and `INSCoreMedia.xcframework` with universal architecture for us 😄

@jessearmandse can you please help regarding this? Need to submit build!

**Comment by jessearmandse** [comment]
The process is not different from the official docs here:
[Creating a multi platform binary framework bundle](https://developer.apple.com/documentation/xcode/creating-a-multi-platform-binary-framework-bundle)

You'd just need to archive the simulator target, then create xcframework from the simulator archived framework and the `arm64` framework.

**Comment by sssuourabh** [comment]
But we don't have the code of INSCameraSDK framework. so how to create an empty target in Xcode with sdk set to iPhonesimulator.?  or what are steps to  ensure your Xcode project has a scheme that builds only the framework target and its dependencies.? @jessearmandse

**Comment by jessearmandse** [comment]
You don't need the code to create xcframework, all you need is the binary.

**Comment by sssuourabh** [comment]
Yes you are correct i am little confused. So my steps should be i create a new project with name of INSCameraSDK with 
<img width="957" alt="Screenshot 2024-07-08 at 1 41 37 PM" src="https://github.com/Insta360Develop/CameraSDK-iOS/assets/2918573/f2d3e667-ee9a-46a5-8dd7-d6e64399bfa9">
and then archive it as mentioned in - https://developer.apple.com/documentation/xcode/creating-a-multi-platform-binary-framework-bundle and then use INSCameraSDK from InstaExample to create a xcframework.?

**Comment by jessearmandse** [comment]
Yup, change the destination appropriately as described there. If the docs are not clear enough, you can refer to this:

[How to create xcframework that contains iOS + iOS Simulator platforms](https://github.com/bielikb/xcframeworks?tab=readme-ov-file#How-to-create-.xcframework-that-contains-iOS-+-iOS-Simulator-platforms)

But, basically the fundamentals are already there in Apple docs

**Comment by sssuourabh** [comment]
okay so do i have to do for both INSCameraSDK ans INSCoreMedia both? @jessearmandse

**Comment by jessearmandse** [comment]
Ahmm, `INSCoreMedia` need a preliminary step to extract the x86_64 architecture, and once you have the `arm64` binary, do the same step as `INSCameraSDK`

Please, read up on why this is the case from Apple official docs for more details, and how I came up with this process. The steps that I did here is not official recommended steps, and depends on your project needs.

Ideally, Insta360 should create this xcframework with both archs. I simply could not wait for them to do so.

**Comment by sssuourabh** [comment]
> target

@jessearmandse i tried this then when i combined and tried build in Simulator in app it says all headers missing from framework. Sometimes it didn't successfully create xcArchives. Please help me if possible, it's urgent. Thanks in advance.

**Comment by sssuourabh** [comment]
HI @jessearmandse i have created xcframework successfully for both INSCameraSDK and INSCoreMedia. Vefore creating xcframework for INSCoreMedia i have extracted the architecture x86_64 from library.

**Comment by wojciechzahradnikdeviniti** [comment]
Hi any updates?

**Comment by Double-Insta360** [comment]
We have already identified this issue in the current version. In future versions, we will default to using XCFramework. If you need it in the near future, please contact Enterprise@insta360.com.

**Comment by jessearmand** [comment]
Also please consider this issue on INSCoreMedia #58
